File: solutions/SIPLinphone/app/src/main/java/ro/pub/cs/systems/eim/lab10/siplinphone/MainActivity.java
package ro.pub.cs.systems.eim.lab10.siplinphone;

import android.Manifest;
import android.content.pm.PackageManager;
import android.os.Bundle;
import android.util.Log;
import android.view.View;
import android.widget.*;
import androidx.appcompat.app.AppCompatActivity;

import org.linphone.core.*;

public class MainActivity extends AppCompatActivity {
    private Core core;

    private final CoreListener coreListener = new CoreListenerStub() {
        @Override
        public void onAccountRegistrationStateChanged(Core core, Account account, RegistrationState state, String message) {
            ((TextView) findViewById(R.id.registration_status)).setText(message);

            if (state != null) {
                switch (state) {
                    case Failed:
                        findViewById(R.id.register).setEnabled(true);
                        break;
                    case Cleared:
                        findViewById(R.id.register_layout).setVisibility(View.VISIBLE);
                        findViewById(R.id.call_layout).setVisibility(View.GONE);
                        findViewById(R.id.register).setEnabled(true);
                        break;
                    case Ok:
                        findViewById(R.id.register_layout).setVisibility(View.GONE);
                        findViewById(R.id.call_layout).setVisibility(View.VISIBLE);
                        findViewById(R.id.unregister).setEnabled(true);
                        findViewById(R.id.remote_address).setEnabled(true);
                        break;
                }
            }
        }

        @Override
        public void onCallStateChanged(Core core, Call call, Call.State state, String message) {
            ((TextView) findViewById(R.id.call_status)).setText(message);

            if (state != null) {
                switch (state) {
                    case IncomingReceived:
                        findViewById(R.id.hang_up).setEnabled(true);
                        findViewById(R.id.answer).setEnabled(true);

                        String remoteAddress = call.getRemoteAddressAsString();
                        if (remoteAddress != null) {
                            ((EditText) findViewById(R.id.remote_address)).setText(remoteAddress);
                        }
                        break;
                    case Connected:
                        findViewById(R.id.mute_mic).setEnabled(true);
                        findViewById(R.id.toggle_speaker).setEnabled(true);
                        Toast.makeText(MainActivity.this, "Remote party answered", Toast.LENGTH_LONG).show();
                        break;
                    case Released:
                        findViewById(R.id.hang_up).setEnabled(false);
                        findViewById(R.id.answer).setEnabled(false);
                        findViewById(R.id.mute_mic).setEnabled(false);
                        findViewById(R.id.toggle_speaker).setEnabled(false);
                        ((EditText) findViewById(R.id.remote_address)).getText().clear();
                        findViewById(R.id.call).setEnabled(true);
                        break;
                }
            }
        }
    };

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.main_activity);

        Factory factory = Factory.instance();
        core = factory.createCore(null, null, this);

        findViewById(R.id.register).setOnClickListener(v -> v.setEnabled(!login()));

        findViewById(R.id.hang_up).setEnabled(false);
        findViewById(R.id.answer).setEnabled(false);
        findViewById(R.id.mute_mic).setEnabled(false);
        findViewById(R.id.toggle_speaker).setEnabled(false);
        findViewById(R.id.remote_address).setEnabled(true);

        findViewById(R.id.answer).setOnClickListener(v -> {
            Call currentCall = core.getCurrentCall();
            if (currentCall != null) {
                currentCall.accept();
            }
        });

        findViewById(R.id.mute_mic).setOnClickListener(v -> {
            core.setMicEnabled(!core.isAgcEnabled());
        });

        findViewById(R.id.toggle_speaker).setOnClickListener(v -> toggleSpeaker());

        findViewById(R.id.call).setOnClickListener(v -> {
            outgoingCall();
            findViewById(R.id.remote_address).setEnabled(false);
            v.setEnabled(false);
            findViewById(R.id.hang_up).setEnabled(true);
        });

        findViewById(R.id.hang_up).setOnClickListener(v -> {
            findViewById(R.id.remote_address).setEnabled(true);
            findViewById(R.id.call).setEnabled(true);

            if (core.getCallsNb() != 0) {
                Call call = core.getCurrentCall() != null ? core.getCurrentCall() : core.getCalls()[0];
                if (call != null) {
                    call.terminate();
                }
            }
        });

        findViewById(R.id.unregister).setOnClickListener(v -> {
            Account account = core.getDefaultAccount();
            if (account != null) {
                AccountParams params = account.getParams();
                AccountParams clonedParams = params.clone();
                clonedParams.setRegisterEnabled(false);
                account.setParams(clonedParams);

                v.setEnabled(false);
            }
        });

        findViewById(R.id.dtmfsend).setOnClickListener(v -> {
            String keypress = ((EditText) findViewById(R.id.dtmfedit)).getText().toString();
            if (keypress.isEmpty()) {
                Toast.makeText(MainActivity.this, "Need phone key character 0-9, +, #", Toast.LENGTH_LONG).show();
                return;
            }

            Call call = core.getCurrentCall() != null
                    ? core.getCurrentCall()
                    : (core.getCalls().length > 0 ? core.getCalls()[0] : null);

            if (call != null) {
                call.sendDtmf(keypress.charAt(0));
            }
        });
    }

    private boolean login() {
        String username = ((EditText) findViewById(R.id.username)).getText().toString();
        String password = ((EditText) findViewById(R.id.password)).getText().toString();
        String domain = ((EditText) findViewById(R.id.domain)).getText().toString();

        TransportType transportType;
        int checkedId = ((RadioGroup) findViewById(R.id.transport)).getCheckedRadioButtonId();
        if (checkedId == R.id.udp) {
            transportType = TransportType.Udp;
        } else if (checkedId == R.id.tcp) {
            transportType = TransportType.Tcp;
        } else {
            transportType = TransportType.Tls;
        }

        Log.i("REGISTER", "success0");
        AuthInfo authInfo = Factory.instance().createAuthInfo(username, null, password, null, null, domain, null);
        Log.i("REGISTER", "success1");

        AccountParams params = core.createAccountParams();
        Address identity = Factory.instance().createAddress("sip:" + username + "@" + domain);
        if (identity == null) {
            Toast.makeText(this, "Identity not valid", Toast.LENGTH_LONG).show();
            return false;
        }
        params.setIdentityAddress(identity);

        Log.i("REGISTER", "success2");
        Address address = Factory.instance().createAddress("sip:" + domain);
        if (address != null) {
            address.setTransport(transportType);
        }
        params.setServerAddress(address);
        params.setRegisterEnabled(true);

        Account account = core.createAccount(params);
        core.addAuthInfo(authInfo);
        core.addAccount(account);

        core.setDefaultAccount(account);
        core.addListener(coreListener);

        core.start();

        if (checkSelfPermission(Manifest.permission.RECORD_AUDIO) != PackageManager.PERMISSION_GRANTED) {
            requestPermissions(new String[]{Manifest.permission.RECORD_AUDIO}, 0);
            return false;
        }
        return true;
    }

    private void toggleSpeaker() {
        AudioDevice currentAudioDevice = core.getCurrentCall() != null
                ? core.getCurrentCall().getOutputAudioDevice()
                : null;
        boolean speakerEnabled = currentAudioDevice != null && currentAudioDevice.getType() == AudioDevice.Type.Speaker;

        for (AudioDevice audioDevice : core.getAudioDevices()) {
            if (speakerEnabled && audioDevice.getType() == AudioDevice.Type.Earpiece) {
                if (core.getCurrentCall() != null) {
                    core.getCurrentCall().setOutputAudioDevice(audioDevice);
                }
                return;
            } else if (!speakerEnabled && audioDevice.getType() == AudioDevice.Type.Speaker) {
                if (core.getCurrentCall() != null) {
                    core.getCurrentCall().setOutputAudioDevice(audioDevice);
                }
                return;
            }
        }
    }

    private void outgoingCall() {
        String remoteSipUri = ((EditText) findViewById(R.id.remote_address)).getText().toString();
        Address remoteAddress = Factory.instance().createAddress("sip:" + remoteSipUri);
        if (remoteAddress == null) {
            return;
        }

        CallParams params = core.createCallParams(null);
        if (params == null) {
            return;
        }

        params.setMediaEncryption(MediaEncryption.None);
        core.inviteAddressWithParams(remoteAddress, params);
    }
}

File: solutions/SIPLinphone/app/src/main/java/ro/pub/cs/systems/eim/lab10/siplinphone/MainActivity.kt

package ro.pub.cs.systems.eim.lab10.siplinphone

import android.Manifest
import android.content.pm.PackageManager
import android.os.Bundle
import android.util.Log
import android.view.View
import android.widget.*
import androidx.appcompat.app.AppCompatActivity
import org.linphone.core.*

class MainActivity:  AppCompatActivity() {
    private lateinit var core: Core

    private val coreListener = object: CoreListenerStub() {
        override fun onAccountRegistrationStateChanged(core: Core, account: Account, state: RegistrationState?, message: String) {
            findViewById<TextView>(R.id.registration_status).text = message

            if (state == RegistrationState.Failed) {
                findViewById<Button>(R.id.register).isEnabled = true
            } else  if (state == RegistrationState.Cleared) {
                findViewById<LinearLayout>(R.id.register_layout).visibility = View.VISIBLE
                findViewById<RelativeLayout>(R.id.call_layout).visibility = View.GONE
                findViewById<Button>(R.id.register).isEnabled = true
            } else if (state == RegistrationState.Ok) {
                findViewById<LinearLayout>(R.id.register_layout).visibility = View.GONE
                findViewById<RelativeLayout>(R.id.call_layout).visibility = View.VISIBLE
                findViewById<Button>(R.id.unregister).isEnabled = true
            }
        }

        override fun onAudioDeviceChanged(core: Core, audioDevice: AudioDevice) {
            // This callback will be triggered when a successful audio device has been changed
        }

        override fun onAudioDevicesListUpdated(core: Core) {
            // This callback will be triggered when the available devices list has changed,
            // for example after a bluetooth headset has been connected/disconnected.
        }

        override fun onCallStateChanged(
            core: Core,
            call: Call,
            state: Call.State?,
            message: String
        ) {
            findViewById<TextView>(R.id.call_status).text = message

            // When a call is received
            when (state) {
                // I N C O M I N G
                Call.State.IncomingReceived -> {
                    findViewById<Button>(R.id.hang_up).isEnabled = true
                    findViewById<Button>(R.id.answer).isEnabled = true

                    val remoteAddress = call.remoteAddressAsString
                    if (remoteAddress != null)
                        findViewById<EditText>(R.id.remote_address).setText(
                            call.remoteAddressAsString ?: "unknown"
                        )
                }
                Call.State.Connected -> {
                    findViewById<Button>(R.id.mute_mic).isEnabled = true
                    findViewById<Button>(R.id.toggle_speaker).isEnabled = true
                }
                Call.State.Released -> {
                    findViewById<Button>(R.id.hang_up).isEnabled = false
                    findViewById<Button>(R.id.answer).isEnabled = false
                    findViewById<Button>(R.id.mute_mic).isEnabled = false
                    findViewById<Button>(R.id.toggle_speaker).isEnabled = false
                    findViewById<EditText>(R.id.remote_address).text.clear()
                    findViewById<Button>(R.id.call).isEnabled = true
                }

                // O U T G O I N G
                Call.State.OutgoingInit -> {
                    // First state an outgoing call will go through
                }
                Call.State.OutgoingProgress -> {
                    // Right after outgoing init
                }
                Call.State.OutgoingRinging -> {
                    // This state will be reached upon reception of the 180 RINGING
                }
                Call.State.Connected -> {
                    // When the 200 OK has been received
                }
                Call.State.StreamsRunning -> {
                    // This state indicates the call is active.
                    // You may reach this state multiple times, for example after a pause/resume
                    // or after the ICE negotiation completes
                    // Wait for the call to be connected before allowing a call update
                }
                Call.State.Paused -> {
                    // When you put a call in pause, it will became Paused
                }
                Call.State.PausedByRemote -> {
                    // When the remote end of the call pauses it, it will be PausedByRemote
                }
                Call.State.Updating -> {
                    // When we request a call update, for example when toggling video
                }
                Call.State.UpdatedByRemote -> {
                    // When the remote requests a call update
                }
                Call.State.Released -> {
                    // Call state will be released shortly after the End state
                    findViewById<EditText>(R.id.remote_address).isEnabled = true
                    findViewById<Button>(R.id.call).isEnabled = true
                    findViewById<Button>(R.id.hang_up).isEnabled = false
                }
                Call.State.Error -> {

                }
            }
        }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        setContentView(R.layout.main_activity)

        val factory = Factory.instance()
        factory.setDebugMode(true, "Hello Linphone")
        core = factory.createCore(null, null, this)

        findViewById<Button>(R.id.register).setOnClickListener {
            login()
            it.isEnabled = false
        }

        findViewById<Button>(R.id.hang_up).isEnabled = false
        findViewById<Button>(R.id.answer).isEnabled = false
        findViewById<Button>(R.id.mute_mic).isEnabled = false
        findViewById<Button>(R.id.toggle_speaker).isEnabled = false
        findViewById<EditText>(R.id.remote_address).isEnabled = true


        findViewById<Button>(R.id.answer).setOnClickListener {
            // if we wanted, we could create a CallParams object
            // and answer using this object to make changes to the call configuration
            // (see OutgoingCall tutorial)
            core.currentCall?.accept()
        }

        findViewById<Button>(R.id.mute_mic).setOnClickListener {
            // The following toggles the microphone, disabling completely / enabling the sound capture
            // from the device microphone
            core.setMicEnabled(!core.isMicEnabled())

        }

        findViewById<Button>(R.id.toggle_speaker).setOnClickListener {
            toggleSpeaker()
        }


        findViewById<Button>(R.id.call).setOnClickListener {
            outgoingCall()
            findViewById<EditText>(R.id.remote_address).isEnabled = false
            it.isEnabled = false
            findViewById<Button>(R.id.hang_up).isEnabled = true
            //findViewById<Button>(R.id.unregister).isEnabled = true

        }

        findViewById<Button>(R.id.hang_up).setOnClickListener {

            findViewById<EditText>(R.id.remote_address).isEnabled = true
            findViewById<Button>(R.id.call).isEnabled = true

            if (core.callsNb != 0) {
                // If the call state isn't paused, we can get it using core.currentCall
                val call = if (core.currentCall != null) core.currentCall else core.calls[0]
                // Terminating a call is quite simple
                if(call != null)
                    call.terminate()
            }
        }

        findViewById<Button>(R.id.unregister).setOnClickListener {
            // Here we will disable the registration of our Account
            val account = core.defaultAccount
            if(account != null) {

                val params = account.params
                // Returned params object is const, so to make changes we first need to clone it
                val clonedParams = params.clone()

                // Now let's make our changes
                clonedParams.setRegisterEnabled(false)

                // And apply them
                account.params = clonedParams

                it.isEnabled = false
            }
        }


    }

    private fun toggleSpeaker() {
        // Get the currently used audio device
        val currentAudioDevice = core.currentCall?.outputAudioDevice
        val speakerEnabled = currentAudioDevice?.type == AudioDevice.Type.Speaker

        // We can get a list of all available audio devices using
        // Note that on tablets for example, there may be no Earpiece device
        for (audioDevice in core.audioDevices) {
            if (speakerEnabled && audioDevice.type == AudioDevice.Type.Earpiece) {
                core.currentCall?.outputAudioDevice = audioDevice
                return
            } else if (!speakerEnabled && audioDevice.type == AudioDevice.Type.Speaker) {
                core.currentCall?.outputAudioDevice = audioDevice
                return
            }/* If we wanted to route the audio to a bluetooth headset
            else if (audioDevice.type == AudioDevice.Type.Bluetooth) {
                core.currentCall?.outputAudioDevice = audioDevice
            }*/
        }
    }


    private fun outgoingCall() {
        // As for everything we need to get the SIP URI of the remote and convert it to an Address
        val remoteSipUri = findViewById<EditText>(R.id.remote_address).text.toString()
        val remoteAddress = Factory.instance().createAddress("sip:" + remoteSipUri)
        remoteAddress ?: return // If address parsing fails, we can't continue with outgoing call process

        // We also need a CallParams object
        // Create call params expects a Call object for incoming calls, but for outgoing we must use null safely
        val params = core.createCallParams(null)
        params ?: return // Same for params

        // We can now configure it
        // Here we ask for no encryption but we could ask for ZRTP/SRTP/DTLS
        params.mediaEncryption = MediaEncryption.None
        // If we wanted to start the call with video directly
        //params.enableVideo(true)

        // Finally we start the call
        core.inviteAddressWithParams(remoteAddress, params)
        // Call process can be followed in onCallStateChanged callback from core listener
    }



    private fun login() {
        val username = findViewById<EditText>(R.id.username).text.toString()
        val password = findViewById<EditText>(R.id.password).text.toString()
        val domain = findViewById<EditText>(R.id.domain).text.toString()
        val transportType = when (findViewById<RadioGroup>(R.id.transport).checkedRadioButtonId) {
            R.id.udp -> TransportType.Udp
            R.id.tcp -> TransportType.Tcp
            else -> TransportType.Tls
        }
        val authInfo = Factory.instance().createAuthInfo(username, null, password, null, null, domain, null)

        val params = core.createAccountParams()
        val identity = Factory.instance().createAddress("sip:$username@$domain")
        params.identityAddress = identity

        val address = Factory.instance().createAddress("sip:$domain")
        address?.transport = transportType
        params.serverAddress = address
        params.setRegisterEnabled(true)
        val account = core.createAccount(params)

        core.addAuthInfo(authInfo)
        core.addAccount(account)

        core.defaultAccount = account
        core.addListener(coreListener)
        core.start()

        // We will need the RECORD_AUDIO permission for video call
        if (packageManager.checkPermission(Manifest.permission.RECORD_AUDIO, packageName) != PackageManager.PERMISSION_GRANTED) {
            requestPermissions(arrayOf(Manifest.permission.RECORD_AUDIO), 0)
            return
        }
    }
}